compiler/rustc_parse/src/parser/token_type.rs

```diff
@@ -0,0 +1,616 @@
+            35 => TokenType::Ident,
+            36 => TokenType::Lifetime,
+            37 => TokenType::Path,
+            38 => TokenType::Type,
+            39 => TokenType::Const,
+
+            40 => TokenType::KwAs,
+            41 => TokenType::KwAsync,
+            42 => TokenType::KwAuto,
+            43 => TokenType::KwAwait,
+            44 => TokenType::KwBecome,
+            45 => TokenType::KwBox,
+            46 => TokenType::KwBreak,
+            47 => TokenType::KwCatch,
+            48 => TokenType::KwConst,
+            49 => TokenType::KwContinue,
+            50 => TokenType::KwCrate,
+            51 => TokenType::KwDefault,
+            52 => TokenType::KwDyn,
+            53 => TokenType::KwElse,
+            54 => TokenType::KwEnum,
+            55 => TokenType::KwExtern,
+            56 => TokenType::KwFn,
+            57 => TokenType::KwFor,
+            58 => TokenType::KwGen,
+            59 => TokenType::KwIf,
+            60 => TokenType::KwImpl,
+            61 => TokenType::KwIn,
+            62 => TokenType::KwLet,
+            63 => TokenType::KwLoop,
+            64 => TokenType::KwMacro,
+            65 => TokenType::KwMacroRules,
+            66 => TokenType::KwMatch,
+            67 => TokenType::KwMod,
+            68 => TokenType::KwMove,
+            69 => TokenType::KwMut,
+            70 => TokenType::KwPub,
+            71 => TokenType::KwRaw,
+            72 => TokenType::KwRef,
+            73 => TokenType::KwReturn,
+            74 => TokenType::KwReuse,
+            75 => TokenType::KwSafe,
+            76 => TokenType::KwSelfUpper,
+            77 => TokenType::KwStatic,
+            78 => TokenType::KwStruct,
+            79 => TokenType::KwTrait,
+            80 => TokenType::KwTry,
+            81 => TokenType::KwType,
+            82 => TokenType::KwUnderscore,
+            83 => TokenType::KwUnsafe,
+            84 => TokenType::KwUse,
+            85 => TokenType::KwWhere,
+            86 => TokenType::KwWhile,
+            87 => TokenType::KwYield,
+
+            88 => TokenType::SymAttSyntax,
+            89 => TokenType::SymClobberAbi,
+            90 => TokenType::SymInlateout,
+            91 => TokenType::SymInout,
+            92 => TokenType::SymIs,
+            93 => TokenType::SymLabel,
+            94 => TokenType::SymLateout,
+            95 => TokenType::SymMayUnwind,
+            96 => TokenType::SymNomem,
+            97 => TokenType::SymNoreturn,
+            98 => TokenType::SymNostack,
+            99 => TokenType::SymOptions,
+            100 => TokenType::SymOut,
+            101 => TokenType::SymPreservesFlags,
+            102 => TokenType::SymPure,
+            103 => TokenType::SymReadonly,
+            104 => TokenType::SymSym,
+
+            _ => panic!("unhandled value: {val}"),
+        };
+        // This assertion will detect if this method and the type definition get out of sync.
+        assert_eq!(token_type as u32, val);
+        token_type
```

Comment:
Can't the function just be the `as` cast with a `<=104` check?

Reply:
No. You can convert a C-style enum to an integer with `as`, but you can't convert in the other direction, e.g. as per [this StackOverflow answer](https://stackoverflow.com/questions/28028854/how-do-i-match-enum-values-with-an-integer/29530566). There are proc macros to do it, but that answer pointed out an alternative that is suitable here: `transmute` is fine so long as the enum is `repr(uN)` for some value of `N`. So I will do that with `repr(u8)`, which will cut over 100 lines of code, yay.

Reply:
Should we have a static assertion that all of them roundtrip? I'm always concerned about a careless future reformat breaking the bidirectional mapping.

Reply:
I'd be happy for extra protection, but I'm having trouble imagining what such a static assertion would look like. Can you explain more?

Reply:
Alternatively, I can go back to an explicit match. The StackOverflow answer mentioned this style:
```
match v {
            x if x == MyEnum::A as i32 => Ok(MyEnum::A),
            x if x == MyEnum::B as i32 => Ok(MyEnum::B),
            x if x == MyEnum::C as i32 => Ok(MyEnum::C),
            _ => Err(()),
        }
```
It requires a line for every variant, but avoids having to write a number on each line.

Reply:
> I'd be happy for extra protection, but I'm having trouble imagining what such a static assertion would look like. Can you explain more?

I'd forgotten that the range operation doesn't work today in const, but I was picturing something like:

```rust
const __CHECK: () = const {
    for i in 0..2 {
        assert_eq!(E::to_i32(&E::from_i32(i)), i);
    }
};
```

Reply:
I've gone with the guard-based version, using a macro to avoid excessive boilerplate. It doesn't rely on `unsafe`, and also doesn't rely on matching up the right integer with the right variant.